### PR TITLE
Add `schema.json` to have a better IDE intellisense

### PR DESCRIPTION
**Pull request by arthurfiorette**
This PR adds an `schema.d.ts` file with all _(documented ones)_ properties in the `turbo.json` file. This way, and combining with the new `npm run jsonschema` script, it's possible to generate an [Json schema](https://json-schema.org/) file.

I configured to the generated file be placed at `docs/public/schema.json` , so it can be requested from the url `https://turborepo.org/schema.json`. You can test how powerfull it is with in IDE autocompletion, intellisense and inline documentation.

_Try it out in your IDE by adding `"$schema": "./docs/public/schema.json"`_

image

It also updated all _(that i saw :P)_ documentation examples to include the `$schema` property, this way, it will go by default for all `Ctrl C/V` users out there.

But i still need to point out some things:

- As the `jsonschema` _(Is that even a good name?)_ script is going to be ran a lot, may we add `ts-json-schema-generator` to the `devDependencies` group?
- In the root `turbo.json` file, i saw an `env` property, but it is not documented, should i add it or it just shouldn't be there?
- I am not a native english speaker, even that i mostly copy pasted the tsdocs in the `schema.d.ts` file, it is not perfect and it probably has some typos.

That's it for now. Thanks for this amazing package!

### CODE REVIEW

**Comment by arthurfiorette** [comment]
https://turbo-site-git-fork-arthurfiorette-main.vercel.sh/schema.json

File: schema.d.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+   * @default origin/master
+   */
+  baseBranch?: string;
+
+  /**
+   * The NPM client in-use in your project.
+   *
+   * @default yarn
+   */
+  npmClient?: "yarn" | "npm" | "pnpm";
```
Comment (jaredpalmer) [inline schema.d.ts:20]: no longer supported / should be removed.

**Comment by arthurfiorette** [comment]
@jaredpalmer anything more?

**Review by pivoamrl — COMMENTED** [review]
thats nice!

File: docs/public/schema.json [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+          "type": "string"
+        },
+        "globalDependencies": {
+          "description": "A list of globs and environment variables for implicit global hash dependencies. Environment variables should be prefixed with $ (e.g. $GITHUB_TOKEN).\n\nAny other entry* without this prefix, will be considered filesystem glob. The contents of these files will be included in the global hashing algorithm and affect the hashes of all tasks.\n\nThis is useful for busting the cache based on .env files (not in Git), environment variables, or any root level file that impacts package tasks (but are not represented in the traditional dependency graph\n\n(e.g. a root tsconfig.json, jest.config.js, .eslintrc, etc.)).",
+          "items": {
+            "type": "string"
+          },
+          "type": "array"
+        },
+        "npmClient": {
```
Comment (jaredpalmer) [inline docs/public/schema.json:53]: Remove

**Review by jaredpalmer — CHANGES_REQUESTED** [review]

File: schema.d.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+   *
+   * @default origin/master
+   */
+  baseBranch?: string;
+
+  /**
+   * A list of globs and environment variables for implicit global hash dependencies.
+   * Environment variables should be prefixed with $ (e.g. $GITHUB_TOKEN).
+   *
+   * Any other entry* without this prefix, will be considered filesystem glob. The
```
Comment (jaredpalmer) [inline schema.d.ts:19]: why is there a *?

Comment (jaredpalmer) [reply schema.d.ts:19]: Let's move this file into `docs`

**Review by jaredpalmer — CHANGES_REQUESTED** [review]
Let's move everything into `docs` folder since this is essentially part of documentation at the moment because it is not directly derived from actual source code

**Review by jaredpalmer — APPROVED** [review]

**Pull request merged**